오픈마켓 [STEP 1] 보리사랑, 웡빙

wongbingg opened the pull request:
안녕하세요 하리보 ! @HARlBO  보리사랑과 웡빙입니다 ! ☺️☺️
@wongbingg , @yusw10
궁금한 사항과 함께 두근두근 첫 PR 을 보냅니다! 정말 기대가 되네요 🤩
# 오픈마켓 STEP1 PR

###### tags: `PR`

## Step 1 기능 구현 : 모델/ 네트워킹 타입 구현
- mock data를 기반으로 `ProductPage` 타입과 `Product` 타입, `Currency` 열거형 구현
- api-host 주소를 저장할 UrlData 열거형 선언
- Json 데이터(Data 타입)를 받아 제네릭T 타입으로 디코딩 해주는 `decode`메서드 구현
- URLSession을 활용하여 네트워크 통신을 하는 `dataTask`메서드 구현
- decode 메서드가 정상동작하는지 mock 데이터를 사용하여 테스트 코드 작성

## UML
### Class Diagram
![](https://i.imgur.com/K6uByTP.png)


## 궁금한 점 / 고민한 점
- Decode, URLSession 에 대한 테스트를 보통 어떻게 진행하는지
    -  Decode에 대해 `test_decode 메서드의 반환값이 Nil 이 아닌지 확인` 만 테스트를 해주었는데, 추가적으로 테스트 할 부분이 있을까요?
    -  URLSession을 활용한 네트워킹을 통해 데이터를 받아왔을 때 그 데이터를 활용하는 것을 테스트 해보려면 특정 키값에 해당하는 밸류값이 잘 들어오는지 확인하는 방법은 어떤가요?
    
- UML 작성시 전역 함수에 대해서 어떻게 표현하는지
  - `decode` 메서드 같은 경우 인스턴스에 넣어서 사용하려면 Json 데이터를 디코딩 할 때마다 인스턴스화 해서 해당 메서드를 사용해야 하는게 불필요하다고 생각해서 현재 전역함수 형태로 구현했습니다. 다만 이렇게 해놓으니 클래스 다이어그램에서 어떻게 표현할지 감이 안와서 일단 추가는 하지 않았습니다. 
  - 전역으로 쓰이는 함수 같은경우에 가능한 방법을 다음과 같이 생각해봤습니다. 
    - 1.전역 함수로 선언하는 방법
    - 2.싱글턴 패턴으로 인스턴스화 해서 사용하는 방법
    - 3.사용할 때 마다 인스턴스화 해주는 법
  - 하리보는 이런 경우에 어떻게 사용하시는지, 그리고 전역함수는 ClassDiagram에 어떻게 표현하시는지 궁금합니다.

yusw10 committed:
feat: "Product" 구조체 및 Currency enum 구현

- 개별 상품 정보를 표현하기 위해 "Product" 데이터 타입 구현
- "Product" 데이터 타입에 Codingkey 및 Codable 프로토콜 채택
- currency 열거형을 구현하여 Product의 currency 프로퍼티에서 사용

wongbingg committed:
feat: "ProductPage" 구조체 구현

- product의 배열을 가지고 있는 상위 모델타입 구현

wongbingg committed:
feat: decodeJsonData() 공용 메서드 생성

- MockData 를 Asset에 추가

yusw10 committed:
test: decodeJsonData메서드의 테스트

- decodeJsonData 메서드가 mock데이터를 디코딩 하는지 확인하는 테스트 추가

yusw10 committed:
feat: URL Session 코드 작성

- 네트워크 통신을 하는 기본 dataTask 메서드 구현

yusw10 committed:
refactor: decode 메서드를 재사용 가능하게 리팩토링

- jsonDecodeData 메서드를 재사용 가능한 형태로 리팩토링 및 메서드명을 decode로 수정
- decode 메서드의 형태에 맞게 dataTask 메서드 리팩토링
- 코드 작성자 주석 수정

yusw10 committed:
test: decode 메서드 테스트 코드 추가

- decode 메서드에 대한 테스트 코드 리팩토링

yusw10 committed:
docs: Class Diagram 추가

- Step1 버전으로 클래스 다이어그램 작성

HARlBO commented on OpenMarket/OpenMarket/Command/URLCommand.swift:
```diff
@@ -0,0 +1,10 @@
+//  Created by 웡빙, 보리사랑 on 2022/07/12.
+//
+
+enum UrlData: String {
```
요기 Url 도 약자이니까  URLData 로 수정하는거 어떨까요!

wongbingg replied:
수정 완료하였습니다!

HARlBO commented on OpenMarket/OpenMarket/Model/Currency.swift:
```diff
@@ -0,0 +1,12 @@
+//
+import Foundation
+
+enum Currency: String, Codable {
```
Currency 를 열거형으로 정의하신 이유를 알고 싶습니닷! 🤔

yusw10 replied:
![스크린샷 2022-07-13 오후 12 10 06](https://user-images.githubusercontent.com/39454792/178642591-2105820d-38c2-4de3-a3fa-5bb43150d5d2.png)

데이터 타입을 정의하는 단계에서 위와 같이 제공된 API 공식문서의 형식을 고려하며 작성을 하려고 노력해보았습니다...!
저는 사실 String 타입이어도 전혀 상관 없다고 생각했는데, 서버에서 enum 타입으로 넘겨주는걸 온전히 받는게 맞다고 생각해서 열거형을 구현하는 방향으로 진행해 보았습니다.

HARlBO commented on OpenMarket/OpenMarket/Model/ProductPage.swift:
```diff
@@ -0,0 +1,37 @@
+        case hasPrev = "has_prev"
+    }
+    
+    func showSelf() {
```
pages 를 출력해주고 있는 함수라서 Self 라는 네이밍보다는 pages 에 관련된 네이밍으로 수정해주면 좋을 것 같아요 🪄

wongbingg replied:
printPage 라는 네이밍으로 수정 해주었습니다 !

HARlBO commented on OpenMarket/OpenMarket/Utility/JsonDecoder.swift:
```diff
@@ -0,0 +1,31 @@
+        case DecodingError.keyNotFound(_ , let context):
+            print(context.debugDescription)
+        default:
+            break
```
case 문에 정의해 놓은 에러가 아닐 경우를 대비해서 `break` 대신
`print(error)` 해주면 좋을 것 같습니닷 👀

wongbingg replied:
더 많은 경우를 고려해야 겠군요! 수정 해주었습니다 ! 😊

HARlBO commented on OpenMarket/OpenMarket/Command/URLCommand.swift:
```diff
@@ -0,0 +1,10 @@
+//
+
+enum UrlData: String {
+    case urlHost = "https://market-training.yagom-academy.kr/"
```
enum 자체가 URLData 라는 의미를 담고 있으니
case 문에서는 `host` 라고만 해줘도 될것 같아요!
그리고 주로 host 는 마지막 / 를 포함하지 않고 path 에 포함해줍니다 🪄

wongbingg replied:
마찬가지로 수정 하였습니다 !

HARlBO commented on OpenMarket/OpenMarket/Utility/URLSession.swift:
```diff
@@ -0,0 +1,40 @@
+
+import Foundation
+
+func dataTask() {
```
클래스나 구조체 안이 아닌 함수만 구현해놓으셨는데 의도하신게 맞을까요?🧐

yusw10 replied:
위에 질문에서 처럼 전역으로 사용되는 메서드의 경우 싱글턴으로 만들어둘지, struct나 class로 감싸서 사용할 때만 인스턴스화 할지 고민해보았습니다. 

- **싱글턴**
  - 장점
    - 불필요한 메모리 낭비를 방지할 수 있다.
  - 단점
    - 코드간 결합도가 높아져서 의존성이 높아진다. 
- **사용할 때 마다 인스턴스화 해서 사용**
  - 장점
    - 사용할 때만 필요한 메서드를 사용이 가능
  - 단점
    - 데이터를 fetch하는 경우가 많을 경우 메모리에 이 메서드를 가진 인스턴스를 계속 생성하는게 불필요하다고 생각함.
- **[현재방식]전역 메서드로 구현해놓고 사용**
  - 장점
    - 코드 어디서나 사용이 가능
  - 단점
    - 접근 제한자를 internal 이상으로 풀어줘야 사용이 가능해서 보안이 취약..?
    
현재는 `dataTask`메서드와 `decode`메서드 두가지를 전역 함수로 만들어 둔 상태입니다. 전역으로 둔 이유는 스텝 진행하면서 빠르게 메서드 동작을 확인하려고 만들었습니다.  인스턴스화 해서 사용하는것과 차이가 있을까요?

HARlBO commented on OpenMarket/OpenMarket/Utility/URLSession.swift:
```diff
@@ -0,0 +1,40 @@
+import Foundation
+
+func dataTask() {
+    let config = URLSessionConfiguration.default
```
접근지정자도 신경써주세요 😉

yusw10 replied:
위에 코멘트처럼 전역으로 사용하려고 internal로 사용해보았습니다.

HARlBO commented on OpenMarket/OpenMarket/Utility/URLSession.swift:
```diff
@@ -0,0 +1,40 @@
+func dataTask() {
+    let config = URLSessionConfiguration.default
+    let session = URLSession(configuration: config)
+    var urlComponents = URLComponents(string: UrlData.urlHost.rawValue + "api/products?")
```
위에 말씀드렸던것 처럼
URLData.host.rawValue + "/api/products"
이런식으로 path 앞에 /를 붙여주면 됩니다!
그리고 path 도 enum 타입으로 관리해주면 더 좋을것 같아요 😉

wongbingg replied:
말씀주신 방법이 더 좋은 것 같아 수정 했습니다! ㅎㅎ 😊

HARlBO commented on OpenMarket/OpenMarket/Utility/URLSession.swift:
```diff
@@ -0,0 +1,40 @@
+    let config = URLSessionConfiguration.default
+    let session = URLSession(configuration: config)
+    var urlComponents = URLComponents(string: UrlData.urlHost.rawValue + "api/products?")
+    let pageNo = URLQueryItem(name: "page_no", value: "1")
```
함수의 재사용성을 위해서 
url, host, queryItem 들을 외부에서 주입받도록 수정하는건 어떨까요?

HARlBO commented on OpenMarket/OpenMarket/Utility/URLSession.swift:
```diff
@@ -0,0 +1,40 @@
+        return
+    }
+    let dataTask = session.dataTask(with: requestURL) { (data, response, error) in
+        guard error == nil else {
```
요 안에 guard 문이 너무 여러개가 나열되어 있어서
합칠 수 있는 부분들은 합친다면 더 깔끔해질것 같습니다

wongbingg replied:
마지막 두 가드문을 묶어서 표현 해주었습니다 !

HARlBO reviewed (commented):
안녕하세요 **웡빙, 보리사랑** 😉
첫 PR 이군요! 이번 프로젝트 같이 화이띵 해봅시다 💪🏻
제가 다는 코멘트들은 정답이 아닐 수 있고, 의도된 바를 제가 파악을 못했을 수도 있답니다!
무조건 수정하지 마시고 언제든 다른 의견들이나 추가적인 질문들 마구마구 해주세요 🙆🏻‍♀️

---

## **궁금한 점 / 고민한 점**

>- Decode, URLSession 에 대한 테스트를 보통 어떻게 진행하는지
>- Decode에 대해 `test_decode 메서드의 반환값이 Nil 이 아닌지 확인` 만 테스트를 해주었는데, 추가적으로 테스트 할 부분이 있을까요?
    - URLSession을 활용한 네트워킹을 통해 데이터를 받아왔을 때 그 데이터를 활용하는 것을 테스트 해보려면 특정 키값에 해당하는 밸류값이 잘 들어오는지 확인하는 방법은 어떤가요?
    
👉🏻 간단한 json mock 데이터를 직접 만들어서, 타입이 다를경우나 값이 없는 경우, 키값이 없는 경우 등    
    지금 decode 함수 catch 문에 정의 해놓으신 에러 타입 정도는 테스트 해주시면 좋을 것 같아요!
    
> - UML 작성시 전역 함수에 대해서 어떻게 표현하는지
>    - `decode` 메서드 같은 경우 인스턴스에 넣어서 사용하려면 Json 데이터를 디코딩 할 때마다 인스턴스화 해서 해당 메서드를 사용해야 하는게 불필요하다고 생각해서 현재 전역함수 형태로 구현했습니다. 다만 이렇게 해놓으니 클래스 다이어그램에서 어떻게 표현할지 감이 안와서 일단 추가는 하지 않았습니다.
>    - 전역으로 쓰이는 함수 같은경우에 가능한 방법을 다음과 같이 생각해봤습니다.
>        - 1.전역 함수로 선언하는 방법
>       - 2.싱글턴 패턴으로 인스턴스화 해서 사용하는 방법
>        - 3.사용할 때 마다 인스턴스화 해주는 법
>    - 하리보는 이런 경우에 어떻게 사용하시는지, 그리고 전역함수는 ClassDiagram에 어떻게 표현하시는지 궁금합니다.
    
  👉🏻 우선 말씀해주신 3가지 방법들 각각 어떤 장단점이 있는지, 그 중에서 전역함수로 선언하는 방법을 선택하신 이유와,
Json 데이터를 디코딩 할때마다 인스턴스화 하는것이 불필요하다고 생각하신 이유를 먼저 들어보고 싶습니다!

yusw10 committed:
test: decode 메서드를 검증하기 위한 테스트 추가

- 타입이 다른 mock data의 추가와 이를 테스트하기위한 테스트코드 작성
- 잘못된 키를 가진 mock data의 추가와 이를 테스트하기위한 테스트코드 작성
- value의 타입이 잘못된 mock data의 추가와 이를 테스트하기위한 테스트코드 작성
- 손상된 mock data의 추가와 이를 테스트하기위한 테스트코드 작성

wongbingg committed:
refactor: 전체적인 네이밍 수정 및 로직변경

- url path 까지 "URLData" enum으로 분리
- DecodingError 의 default 인 경우도 error를 출력하도록 변경
- guard let 구문 결합

HARlBO reviewed (approved):
말씀하신 것 처럼 인스턴스화 해서 사용하게 되면 인스턴스를 생성할때 메모리에 올라갔다가 사용이 끝나면 메모리에서 해제 되게 됩니다. 

반면 전역함수는 프로그램 시작과 함께 메모리에 올라가서 함수를 사용하지 않아도 프로그램이 끝날때까지 메모리 할당이 되어 있어요!
그치만 이 비용보다 특정 객체를 생성하는 비용이 더 많다고 판단이 되거나 전역 함수처럼 공용으로 사용하는 느낌일때, 이런 상황에서는 타입 메소드도 많이 사용한답니다! 
전역 함수는 질문 주신 것 처럼 UML 에서 표현도 어렵고, 프로그램 전체에 관련된 함수인지 역할이 명확하지 않아서 특정 클래스의 타입처럼 사용하는게 좋을 수 있어요 🪄

두분이서 좀 더 생각해보시고 변경사항이 있다면 다음 스텝에서 반영해주세요 😉
STPE 3 에서 만나요 🏃🏻‍♂️
